clang/runtime/dpct-rt/include/dpl_extras/algorithm.h

```diff
@@ -1455,8 +1455,604 @@ inline void segmented_sort_pairs_by_two_pair_sorts(
+                    bins_buf.all_view().end(), __histo_value_type(0));
+  auto N = last - first;
+  if (num_bins < __max_registers) {
+    // If bins fit into registers, use register private accumulation
+    __histogram_general_registers_local_reduction<1024, 32, 16,
+                                                  __histo_value_type>(
+        ::std::forward<Policy>(policy), input_buf.all_view(),
+        bins_buf.all_view(), num_bins, func,
+        std::forward<Range...>(opt_range)...);
+  } else if (num_bins * sizeof(__histo_value_type) < __local_mem_size) {
+    // If bins fit into SLM, use local atomics
+
+    // Experimentally determined iters per work-item
+    if (N <= 524288) {
+      __histogram_general_local_atomics<1024, 4, __histo_value_type>(
+          ::std::forward<Policy>(policy), input_buf.all_view(),
+          bins_buf.all_view(), num_bins, func,
+          std::forward<Range...>(opt_range)...);
+    } else {
+      __histogram_general_local_atomics<1024, 32, __histo_value_type>(
+          ::std::forward<Policy>(policy), input_buf.all_view(),
+          bins_buf.all_view(), num_bins, func,
+          std::forward<Range...>(opt_range)...);
+    }
+  } else // Otherwise, use global atomics (private copies per workgroup)
+  {
+    // Experimentally determined iters per work-item
+    if (N <= 524288) {
+      __histogram_general_private_global_atomics<1024, 4, __histo_value_type>(
+          ::std::forward<Policy>(policy), input_buf.all_view(),
+          bins_buf.all_view(), num_bins, func,
+          std::forward<Range...>(opt_range)...);
+    } else {
+      __histogram_general_private_global_atomics<1024, 32, __histo_value_type>(
+          ::std::forward<Policy>(policy), input_buf.all_view(),
+          bins_buf.all_view(), num_bins, func,
+          std::forward<Range...>(opt_range)...);
+    }
+  }
+  return histogram_first + num_bins;
+}
+
+template <typename Policy, typename Iter1, typename Iter2, typename Size,
+          typename T>
+inline ::std::enable_if_t<
+    dpct::internal::is_iterator<Iter1>::value &&
+        dpct::internal::is_iterator<Iter2>::value &&
+        internal::is_hetero_execution_policy<::std::decay_t<Policy>>::value,
+    Iter2>
+histogram(Policy &&policy, Iter1 first, Iter1 last, Iter2 histogram_first,
+          const Size &num_bins, const T &first_bin_min_val,
+          const T &last_bin_max_val) {
+  return __histogram_general_select_best(
+      ::std::forward<Policy>(policy), first, last, histogram_first, num_bins,
+      __evenly_divided_binhash<T>(first_bin_min_val, last_bin_max_val,
+                                  num_bins));
+}
+
+template <typename Policy, typename Iter1, typename Iter2, typename Iter3>
+inline ::std::enable_if_t<
+    dpct::internal::is_iterator<Iter1>::value &&
+        dpct::internal::is_iterator<Iter2>::value &&
+        dpct::internal::is_iterator<Iter3>::value &&
+        internal::is_hetero_execution_policy<::std::decay_t<Policy>>::value,
```

Comment:
Do we currently plan to support a host policy version? This may be more relevant for when putting into oneDPL.

Reply:
There will be a plan to have this in oneDPL, but perhaps not initially.  We do not plan to support a host policy version in this PR.  However, once it is in oneDPL, it could be supported here.